core/src/main/java/org/jruby/ext/thread/Mutex.java

```diff
@@ -92,7 +92,10 @@ public IRubyObject lock(ThreadContext context) {
         try {
             thread.enterSleep();
             checkRelocking(context);
-            thread.lock(lock);
+            thread.lockInterruptibly(lock);
+        } catch (InterruptedException ie) {
+            context.pollThreadEvents();
+            throw context.runtime.newConcurrencyError("interrupted waiting for mutex");
```

Comment:
Interrupt is probably not right because it typically corresponds to SIGINT and usually leads to termination of the runtime. This patch basically undoes a "fix" I did in #4261. Simply going back to interruptible/Interrupted exception would undo that "fix" and break LogStash. We need a better option.